src/main/java/hudson/tasks/MimeMessageBuilder.java

```diff
@@ -0,0 +1,198 @@
+        return this;
+    }
+
+    public MimeMessageBuilder setDefaultSuffix(String defaultSuffix) {
+        assert defaultSuffix != null;
+        this.defaultSuffix = defaultSuffix;
+        return this;
+    }
+
+    public MimeMessageBuilder setFrom(String from) {
+        assert from != null;
+        this.from = from;
+        return this;
+    }
+
+    public MimeMessageBuilder setReplyTo(String replyTo) {
+        assert replyTo != null;
+        this.replyTo = replyTo;
+        return this;
+    }
+
+    public MimeMessageBuilder setRecipients(String recipients) {
+        assert recipients != null;
+        this.recipients = recipients;
+        return this;
+    }
+
+    /**
+     * Build a {@link MimeMessage} instance from the set of supplied parameters.
+     * @return The {@link MimeMessage} instance;
+     * @throws MessagingException
+     * @throws UnsupportedEncodingException
+     */
+    public MimeMessage buildMimeMessage() throws MessagingException, UnsupportedEncodingException {
+        MimeMessage msg = new MimeMessage(Mailer.descriptor().createSession());
+
+        msg.setContent("", mimeType);
+        if (StringUtils.isNotBlank(from)) {
+            msg.setFrom(toNormalizedAddress(from));
+        }
+        msg.setSentDate(new Date());
+
+        addRunHeaders(msg);
+        addRecipients(msg);
+
+        if (StringUtils.isNotBlank(replyTo)) {
+            msg.setReplyTo(new Address[]{toNormalizedAddress(replyTo)});
+        }
+        return msg;
+    }
+
+    public InternetAddress toNormalizedAddress(String address) throws UnsupportedEncodingException {
+        if (address == null) {
+            return null;
+        }
+
+        // if not a valid address (i.e. no '@'), then try adding suffix
+        if (!address.contains("@")) {
+            if (defaultSuffix != null && defaultSuffix.contains("@")) {
+                address += defaultSuffix;
+            } else {
+                return null;
+            }
+        }
+
+        try {
+            return Mailer.StringToAddress(address, charset);
+        } catch (AddressException e) {
+            // report bad address, but try to send to other addresses
+            if (listener != null) {
+                listener.getLogger().println("Unable to send to address: " + address);
+                e.printStackTrace(listener.error(e.getMessage()));
+            }
+            return null;
+        }
+    }
+
+    public static void setInReplyTo(MimeMessage msg, String inReplyTo) throws MessagingException {
+        if (msg == null || inReplyTo == null) {
+            return;
+        }
+        msg.setHeader("In-Reply-To", inReplyTo);
+        msg.setHeader("References", inReplyTo);
+    }
+
+    private void addRunHeaders(MimeMessage msg) throws MessagingException {
+        if (run != null) {
+            msg.addHeader(X_JENKINS_JOB, run.getParent().getFullName());
+            msg.addHeader(X_JENKINS_RESULT, run.getResult().toString());
```

Comment:
This seems specific to `MailSender`. I see no reason why it should be in the API.